tools/mockgen/mockgen.go

```diff
@@ -0,0 +1,102 @@
+// Package mockgen contains a wrapper for the `mockgen` tool, which bypasses running `mockgen` if the destination file
+// is newer than the source file. This is useful because `mockgen` is slow and can be a bottleneck in the build process.
+package mockgen
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+// Run a cached version of `mockgen`, which checks the modification times of the source and destination files,
+// running mockgen only if necessary. This is similar to the behavior of `Make`, but it wasn't easy to express this
+// in a Makefile because the generate commands are in our source code.
+func Run(args []string, opts ...Option) error {
+	params := &Params{
+		Exec: runRealCommand,
+	}
+	for _, opt := range opts {
+		opt(params)
+	}
+
+	upToDate, err := isDestinationFileUpToDateWithSourceInArgs(args)
+	if err != nil {
+		return err
+	}
+	if !upToDate {
+		return nil
+	}
+
+	// We couldn't confirm that the target file was up-to-date; run mockgen
+	return params.Exec(args)
+}
+
+// Params for running the mockgen wrapper.
+type Params struct {
+	// Exec is a function that should execute a shell command with the given arguments.
+	Exec func(args []string) error
+}
+
+// Option to override the [Params] of the mockgen wrapper.
+type Option func(*Params)
+
+// WithExecFn sets a custom [Params.Exec] function.
+func WithExecFn(f func(args []string) error) Option {
+	return func(params *Params) {
+		params.Exec = f
+	}
+}
+
+func runRealCommand(args []string) error {
+	fullPath, lookErr := exec.LookPath("mockgen")
+	if lookErr != nil {
+		return fmt.Errorf("failed to find mockgen binary: %w", lookErr)
+	}
+	err := syscall.Exec(fullPath, append([]string{""}, args...), os.Environ())
+	if err != nil {
+		return fmt.Errorf("failed to exec mockgen with args %v: %w", args, err)
+	}
+	return nil
```

Comment:
I think this can be a lot simpler?
```suggestion
	return exec.Command("mockgen", args...).Run()
```

Reply:
I get that exec is nicer in theory in some ways, but.. does it matter here? I'd rather see simpler code than save 100µs